File: pdf_scraper.py
import os
import re
import requests
import concurrent.futures
from bs4 import BeautifulSoup

BASE_URL = "https://papers.nips.cc/paper/"
DOWNLOAD_DIR = "mydocuments"
THREAD_POOL_SIZE = 30  # Adjust based on system capability


def create_download_dir():
    """Create download directory if not exists."""
    if not os.path.exists(DOWNLOAD_DIR):
        os.makedirs(DOWNLOAD_DIR)


def get_paper_links(year):
    """Scrapes paper links from the given year."""
    url = f"{BASE_URL}{year}"
    print(f"\nScraping year: {year}")

    response = requests.get(url)
    if response.status_code != 200:
        print(f"Failed to retrieve data for {year}")
        return []

    soup = BeautifulSoup(response.text, 'html.parser')
    paper_elements = soup.select("li.conference a[href], li.none a[href]")

    paper_links = []
    for paper in paper_elements:
        paper_title = paper.text
        paper_url = "https://papers.nips.cc" + paper['href']
        paper_links.append((paper_title, paper_url, year))

    return paper_links


def download_paper(paper_title, paper_url, year):
    """Processes and downloads the PDF of a paper."""
    print(f"\nProcessing: {paper_title}")

    response = requests.get(paper_url)
    if response.status_code != 200:
        print(f"Failed to fetch page: {paper_title}")
        return False

    soup = BeautifulSoup(response.text, 'html.parser')
    pdf_links = soup.select("a[href$='.pdf']")
    
    if not pdf_links:
        print(f"No valid document found for: {paper_title}")
        return False

    for idx, pdf_link in enumerate(pdf_links):
        pdf_url = "https://papers.nips.cc" + pdf_link['href']
        file_suffix = "_1.pdf" if idx == 0 else f"_{idx+1}.pdf"
        download_file(pdf_url, paper_title, file_suffix)

    return True


def download_file(url, title, suffix):
    """Downloads a file from a given URL."""
    safe_title = re.sub(r"[^a-zA-Z0-9.-]", "_", title) + suffix
    file_path = os.path.join(DOWNLOAD_DIR, safe_title)

    response = requests.get(url, stream=True)
    if response.status_code == 200:
        with open(file_path, "wb") as f:
            for chunk in response.iter_content(1024):
                f.write(chunk)
        print(f"Downloaded: {safe_title}")
    else:
        print(f"Failed to download: {safe_title}")


def main():
    create_download_dir()
    year = 2021  # Adjust the year as needed
    paper_links = get_paper_links(year)

    with concurrent.futures.ThreadPoolExecutor(max_workers=THREAD_POOL_SIZE) as executor:
        futures = [executor.submit(download_paper, title, url, year) for title, url, year in paper_links]
        concurrent.futures.wait(futures)

    print("\nAll downloads complete!")


if __name__ == "__main__":
    main()

File: PdfScraper.java
package scrap;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.apache.hc.client5.http.classic.methods.HttpGet;
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
import org.apache.hc.client5.http.impl.classic.HttpClients;
import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.util.Set;
import java.util.concurrent.*;

public class PdfScraper {

    private static final String BASE_URL = "https://papers.nips.cc/paper/";
    private static final String DOWNLOAD_DIR = "mydocuments";
    private static final String[] FILE_EXTENSIONS = {".pdf", ".docx", ".txt", ".avis"};
    private static final int THREAD_POOL_SIZE = 30; // Adjust based on system capability

    private static final ConcurrentHashMap<String, Boolean> completedDownloads = new ConcurrentHashMap<>();

    public static void main(String[] args) {
        ExecutorService executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE);

        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
            File downloadDir = new File(DOWNLOAD_DIR);
            if (!downloadDir.exists()) {
                downloadDir.mkdir();
            }

            for (int year = 2021; year >= 2021; year--) {
                final int currentYear = year; // Create a final copy of the year variable
                String yearUrl = BASE_URL + currentYear;
                System.out.println("\nScraping year: " + currentYear);

                Document document = Jsoup.connect(yearUrl).get();
                Elements paperElements = document.select("li.conference a[href], li.none a[href]");

                for (Element paperElement : paperElements) {
                    String paperTitle = paperElement.text();
                    String paperUrl = "https://papers.nips.cc" + paperElement.attr("href");

                    // Use the final copy of the year variable in the lambda
                    executor.submit(() -> processPaper(httpClient, paperTitle, paperUrl, currentYear));
                }
            }

            // Shutdown the executor and wait for all tasks to complete
            executor.shutdown();
            executor.awaitTermination(30, TimeUnit.MINUTES);

            // Print a summary of completed downloads
            System.out.println("\nDownload Summary:");
            completedDownloads.forEach((file, status) -> 
                System.out.println(file + " -> " + (status ? "Success" : "Failed"))
            );

            System.out.println("\nAll downloads complete!");

        } catch (IOException | InterruptedException e) {
            e.printStackTrace();
        }
    }

    private static void processPaper(CloseableHttpClient httpClient, String paperTitle, String paperUrl, int year) {
        try {
            System.out.println("\nProcessing: " + paperTitle);
            Document paperDocument = Jsoup.connect(paperUrl).get();
            boolean downloaded = false;

            Elements pdfLinks = paperDocument.select("a[href$=.pdf]");
            if (!pdfLinks.isEmpty()) {
                // Iterate over all PDF links and download each one
                for (int i = 0; i < pdfLinks.size(); i++) {
                    Element pdfLink = pdfLinks.get(i);
                    if (pdfLink != null) {
                        String fileUrl = "https://papers.nips.cc" + pdfLink.attr("href");
                        String fileSuffix = i == 0 ? "_1.pdf" : "_2.pdf"; // Add suffix to differentiate files
                        if (downloadFile(httpClient, fileUrl, paperTitle, fileSuffix)) {
                            completedDownloads.put(paperTitle + fileSuffix, true);
                            downloaded = true;
                        }
                    }
                }
            }

            if (!downloaded) {
                System.out.println("No valid document found for: " + paperTitle);
                completedDownloads.put(paperTitle, false);
            }
        } catch (IOException e) {
            System.out.println("Failed to process: " + paperTitle);
            completedDownloads.put(paperTitle, false);
        }
    }

    private static boolean downloadFile(CloseableHttpClient httpClient, String fileUrl, String paperTitle, String fileSuffix) {
        try {
            String fileName = paperTitle.replaceAll("[^a-zA-Z0-9.-]", "_") + fileSuffix;
            File outputFile = new File(DOWNLOAD_DIR, fileName);

            HttpGet request = new HttpGet(fileUrl);
            try (CloseableHttpResponse response = httpClient.execute(request)) {
                if (response.getEntity() != null) {
                    FileUtils.copyInputStreamToFile(response.getEntity().getContent(), outputFile);
                    System.out.println("Downloaded: " + fileName);
                    return true;
                } else {
                    System.out.println("No content received for: " + paperTitle);
                }
            }
        } catch (IOException e) {
            System.out.println("Failed to download: " + paperTitle);
        }
        return false;
    }
}